[libc] Propagate printf config options from a single config header library.

sivachandra opened the pull request:
printf_core.parser is not yet updated to use the printf config options; It
does not use them currently anyway and the corresponding parser_test
should be updated to respect the config options.

michaelrj-google commented on libc/src/stdio/printf_core/CMakeLists.txt:
```diff
@@ -1,3 +1,25 @@
+if(LIBC_CONF_PRINTF_DISABLE_FLOAT)
```
if these conditionals are being moved into `printf_core` should the `add_subdirectory` lines in stdio be moved back up to the top?

sivachandra replied:
Where do `add_subdirectory` in `stdio` becomes irrelevant from the config options point of view - it should be at a place where it makes the most practical sense.

michaelrj-google commented on libc/src/stdio/CMakeLists.txt:
```diff
@@ -188,27 +188,14 @@ list(APPEND printf_deps
   )
 else()
-  list(APPEND printf_copts "-DLIBC_COPT_PRINTF_USE_SYSTEM_FILE")
+  list(APPEND use_system_file "COMPILE_OPTIONS" "-DLIBC_COPT_PRINTF_USE_SYSTEM_FILE")
```
is there a benefit to moving the `COMPILE_OPTIONS` tag into the option? Also, I'm not sure this name is the best option since this option will only work for printf, but scanf will need a similar one.

sivachandra replied:
Yes - when there are no `COMPILE_OPTIONS`, then CMake will not treat whatever follows `COMPILE_OPTIONS` as an intended compile option.

About the name, do you mean `LIBC_COPT_PRINTF_USE_SYSTEM_FILE`? If yes, then you can either change it when you add similar feature to `scanf`.  If you are talking about `use_system_file`, you can share it with `scanf` targets, or add a new one at that point.

Commit:
[libc] Propagate printf config options from a single config header library.

printf_core.parser is not yet updated to use the printf config options; It
does not use them currently anyway and the corresponding parser_test
should be updated to respect the config options.

sivachandra commented:
Rebased. @jhuber6 - Can you help with testing this on the GPU?

jhuber6 commented:
I'm getting
```
ld.lld: error: unable to find library -llibc.include.gpu_rpc.__generated_hdr__
```
Any clue what that could be? This seems to work for NVPTX, presumably because we don't use static libraries and everything is an object library.

Commit:
[libc] Add add_gen_header deps separately.

sivachandra commented:
> I'm getting
> 
> ```
> ld.lld: error: unable to find library -llibc.include.gpu_rpc.__generated_hdr__
> ```
> 
> Any clue what that could be? This seems to work for NVPTX, presumably because we don't use static libraries and everything is an object library.

Can you give it a try now? I hoping the second commit should have fixed it.

jhuber6 reviewed (approved):
Passes now, thanks.
